walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,206 @@
+                }
+
+                if (TimeRange.END_OF_DAY - time.end() >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(time.end(), TimeRange.END_OF_DAY, true));
+                }
+            } else {
+                times.add(TimeRange.WHOLE_DAY);
+            }
+
+            return times;
+        } else {
+            Event[] e = events.toArray(new Event[0]);
+            Event[] sortedEvents = insertionSort(e, TimeRange.ORDER_BY_START, request);;
+
+            if (sortedEvents.length == 0) {
+                if (optionalEvents.size() == 0) {
+                    times.add(TimeRange.WHOLE_DAY);
+                }
+        
+                return times;
+            } else if (sortedEvents.length == 1) {
+                TimeRange time = sortedEvents[0].getWhen();
+
+                if (time.start() - TimeRange.START_OF_DAY >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, time.start(), false));
+                }
+
+                if (TimeRange.END_OF_DAY - time.end() >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(time.end(), TimeRange.END_OF_DAY, true));
+                }
+
+                return times;
+            } else {
+                int startTime = TimeRange.START_OF_DAY;
+                int endTime = TimeRange.END_OF_DAY;
+
+                TimeRange first = sortedEvents[0].getWhen();
+                TimeRange second = sortedEvents[1].getWhen();
+
+                int tempStart = startTime;
+                int tempEnd = endTime;
+
+                boolean optional = false;
+                int previous = startTime;
+
+                if (first.overlaps(second)) {
+                    tempEnd = Math.min(first.start(), second.start());
+                    tempStart = Math.max(first.end(), second.end());
+
+                    if (optionalEvents.contains(first)) {
+                        if (first.start() - startTime < (int) request.getDuration()) {
+                            tempEnd = second.start();
+                            optional = true;
+                            tempStart = previous;
+                        }
+                    }
+                } else {
+                    if (optionalEvents.contains(first)) {
+                        if (first.start() - startTime < (int) request.getDuration()) {
+                            tempEnd = second.start();
+                            optional = true;
+                        }
+                    } else {
+                        tempEnd = first.start();
+                        previous = first.end();
+                    }
+                }
+
+                if (tempEnd - startTime >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(startTime, tempEnd, false));
+                }
+
+                for (int i = 0; (i + 1) < sortedEvents.length; i++) {
```

Comment:
You have already examined sortedEvents[0] and sortedEvents[1], so why does this loop need to start at i=0?

Conversely, why do you need special handling of the first two events?

Reply:
Prior to the execution of the loop, the gap between the beginning of the day and the first event (and second event if there is overlap) is evaluated. Inside of the loop however, the gap between the first and second events is evaluated with another final evaluation going over the gap between the last event and the end of the day.